### PR TITLE
WL21387 - Avatar Bookmarks now save Attachment Data and scale

### PR DESCRIPTION
**Pull request by Menithal**
Old avatar and location bookmarks should still work.
To test Access the Avatar menu and save / load / delete avatars. This should save an avatarbookmarks.json file under Roaming Appdata.

Also test Location Bookmarks again, as this update refactored the entire Bookmark system lightly.

Changes:

- Now Saves Attachments with Avatar Bookmark
- Now Saves Scale with Avatar Bookmark
- Added MyAvatar.getAvatarScale and  MyAvatar.setAvatarScale for programatic and script access to getTargetScale and setTargetScale / scale.

Test the following:

- Save Avatar Bookmark
- Load Avatar Bookmark
- Delete Avatar Bookmark
- Access MyAvatar.getAvatarScale() and see value being same as MyAvatar.scale
- MyAvatar.setAvatarScale(float) and mirror with MyAvatar.scale = float
- Save Place Bookmark
- Load Place Bookmark
- Delete Place Bookmark

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1093/HighFidelity-Beta-PR10816-359642eb0e2fa4e434a5b6bc7919ab9ab2f21597.exe

**Comment by CainFoool** [comment]
Hey @Menithal; is this ready?

**Comment by Menithal** [comment]
@CainFoool apparently there was some build failure and I'd need read to hifi project in jenkins to see what went wrong. so not yet.

**Comment by CainFoool** [comment]
@Menithal: That happens often at the minute with the latest builds, Windows was okay - Mac was broken. It more than likely just needs to be rebuilt.

**Comment by Menithal** [comment]
If that's so feel free to test, but this does touch c++ code it may effect the Mac code as well.

**Comment by CainFoool** [comment]
Alright - I'll give it a go. Testing.

**Comment by CainFoool** [comment]
# Test Failed 
Test | Result
-------- | --------
Save Avatar Bookmark | Fail
Load Avatar Bookmark | Fail
Delete Avatar Bookmark | Fail
Access MyAvatar.getAvatarScale() and see value being same as MyAvatar.scale | Pass
MyAvatar.setAvatarScale(float) and mirror with MyAvatar.scale = float | Fail
Save Place Bookmark | Pass
Load Place Bookmark | Pass
Delete Place Bookmark | Pass

## Avatar Bookmarks
This error occured when saving and deleting any bookmarks. Additionally, it doesn't look like the Interface does a check when it loads. The "Avatar Bookmarks" ComboBox is enabled regardless of there being bookmarks stored or not - only when you delete an Avatar Bookmark does it disable. Running it in Administrator mode didn't resolve the issue either for the saving/deleting issue. Loading an Avatar didn't appear to do anything.
```
[06/26 09:30:26] [DEBUG] [hifi.interface] Deleted bookmark: "Test"
[06/26 09:30:26] [WARNING] [default] Couldn't open bookmarks file for writing
```

Loading an Avatar Bookmark:
```
[06/26 09:40:20] [DEBUG] [hifi.interface] Avatar On  "file:///C:/Program Files/High Fidelity - PR10816/resources/meshes/defaultAvatar_full.fst"
[06/26 09:40:20] [DEBUG] [hifi.interface] Attach  ()
```

## Avatar Scaling
See below for console log. setAvatarScale(float); appears to be returning undefined and not actually setting the Avatar Scale.
```
MyAvatar.getAvatarScale();
1
MyAvatar.scale;
1
MyAvatar.setAvatarScale(2);
undefined
MyAvatar.scale;
1
MyAvatar.scale = 2;
2
MyAvatar.scale;
2
```

Let me know once you'd like a retest.
###### Intel i5 7600K @ 4Ghz, 16GB DDR4 2133Mhz Corsair Vengeance, Asus Strix GTX 1050Ti 3GB, HTC Vive

**Comment by Menithal** [comment]
As a side note, functions if they return nothing will always result in undefined, and this is common for all Setters.

Here is a question Did you already have a pre-existing avatar bookmark (as in, does the file already exist in `AppdataPath High Fidelity PR / Roaming/Interface/` ? or was from completely scratch?

also the other question is, did setAvatarScale scale your avatar at all, even without referring to scale?

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -2524,6 +2524,21 @@ bool MyAvatar::getFlyingEnabled() {
...
 
+// Public interface for targetscale
+float MyAvatar::getAvatarScale() {
+    return getTargetScale();
+}
+
+void MyAvatar::setAvatarScale(float val) {
+
+    if (QThread::currentThread() != thread()) {
+        QMetaObject::invokeMethod(this, "setScale", Q_ARG(float, val));
```
Comment (Menithal) [inline interface/src/avatar/MyAvatar.cpp]: Need to fix this one, should be setAvatarScale, not setScale.

**Comment by CainFoool** [comment]
> Did you already have a pre-existing avatar bookmark (as in, does the file already exist in AppdataPath High Fidelity PR / Roaming/Interface/ ? or was from completely scratch? What happens if you transfer a bookmark from the stable, forexample.

This was a brand new installation with no bookmarks copied from stable. Let me do a quick retest with a bookmark from Stable and get back to you on that one. 

> did setAvatarScale scale your avatar at all, even without referring to scale?
setAvatarScale did nothing, .scale was the only function which did anything.

**Comment by Menithal** [comment]
Ok, might have something to do with file creation, ill look into it, but get back to me on if the copying over worked.

**Comment by CainFoool** [comment]
@Menithal:

Looks like I might be on to something. In Master, the same error message is occuring when saving a Bookmark, and again that's with running it as Admin (Just to ensure permissions weren't the issue here); so the issue isn't relevant to your PR.

```
[06/26 10:16:08] [DEBUG] [hifi.interface] Added bookmark: "defaultAvatarGuitarThingThroughHead" , "file:///C:/Program Files/High Fidelity/resources/meshes/defaultAvatar_full.fst"
[06/26 10:16:08] [WARNING] [default] Couldn't open bookmarks file for writing
```

**Comment by Menithal** [comment]
@CainFoool OK,
I was able to replicate the avatar scaling and fixed that.

Everything else, unfortunately I cannot replicate on two of my machines. You have some odd issue with permissions perhaps?

**Comment by CainFoool** [comment]
@Menithal: I'll try messing with the folder permissions, it's completely possible that something messed up somewhere on my side. Though it looks like your latest build failed completely this time.

**Comment by Menithal** [comment]
@CainFoool Going to close this PR and Restart it, apparently there is an issue with jenkins.

**Pull request closed without merging**